### PR TITLE
refactor(connector): [Klarna] Add support for Klarna Optional Shipping Address

### PR DESCRIPTION
**Pull request by swangi-kumari**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
1. Create a Merchant 
```
{
  "merchant_id": "merchant_{{$timestamp}}",
  "locker_id": "m0010",
  "merchant_name": "NewAge Retailer",
  "merchant_details": {
    "primary_contact_person": "John Test",
    "primary_email": "JohnTest@test.com",
    "primary_phone": "sunt laborum",
    "secondary_contact_person": "John Test2",
    "secondary_email": "JohnTest2@test.com",
    "secondary_phone": "cillum do dolor id",
    "website": "https://www.example.com",
    "about_business": "Online Retail with a wide selection of organic products for North America",
    "address": {
      "line1": "1467",
      "line2": "Harrison Street",
      "line3": "Harrison Street",
      "city": "San Fransico",
      "state": "California",
      "zip": "94122",
      "country": "US"
    }
  },
  "return_url": "https://google.com/success",
  "webhook_details": {
    "webhook_version": "1.0.1",
    "webhook_username": "ekart_retail",
    "webhook_password": "password_ekart@123",
    "payment_created_enabled": true,
    "payment_succeeded_enabled": true,
    "payment_failed_enabled": true
  },
  "sub_merchants_enabled": false,
  "metadata": {
    "city": "NY",
    "unit": "245"
  },
  "primary_business_details": [
    {
      "country": "US",
      "business": "food"
    }
  ]
}
```
___________________________
2. Create API Key 
```
{
  "name": "API Key 1",
  "description": null,
  "expiration": "2038-01-19T03:14:08.000Z"
}
```
_________________________________
3. Create MCA
```
{
    "connector_type": "fiz_operations",
    "connector_name": "klarna",
    "connector_account_details": {
        "auth_type": "BodyKey",
        "api_key": "Password",
        "key1": "username"
    },
    "test_mode": true,
    "disabled": false,
    "payment_methods_enabled": [
        {
            "payment_method": "pay_later",
            "payment_method_types": [
                {
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true,
                    "payment_experience": "invoke_sdk_client",
                    "payment_method_type": "klarna"
                }
            ]
        },
        {
            "payment_method": "pay_later",
            "payment_method_types": [
                {
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true,
                    "payment_experience": "redirect_to_url",
                    "payment_method_type": "klarna"
                }
            ]
        }
    ],
    "metadata": {
        "city": "NY",
        "unit": "245",
        "klarna_region": "NorthAmerica"
    }
}
```
  _______________________________________________________________
  Create a payment via sdk with capture_method = `automatic` and use the shipping address which is provided below
```
{
      currency: "USD",
      amount: 5000,
      order_details: [
        {
          product_name: "Apple iphone 15",
          quantity: 1,
          amount: 5000,
        },
      ],
      confirm: false,
      capture_method: "automatic",
      authentication_type: "three_ds",
      customer_id: "hyperswitch_sdk_demo_id",
      email: "hyperswitch_sdk_demo_id@gmail.com",
      description: "Hello this is description",
      connector_metadata: {
        noon: {
          order_category: "applepay",
        },
      },
      metadata: {
        udf1: "value1",
        new_customer: "true",
        login_date: "2019-09-10T10:11:12Z",
      },
      shipping: {
        address: {
          state: "NY",
          city: "New York",
          country: "US",
          line1: "509",
          line2: "Amsterdam Ave",
          line3: "alsksoe",
          zip: "10024",
          first_name: "Test",
          last_name: "Person-us",
        },
        phone: {
          number: "5551234",
          country_code: "+1",
        },
        email: "swangi@gmail.com"
      },
      billing: {
        address: {
          country: "US",
        }
      },
      }
```
Check in klarna dashboard if shipping details are being populated correctly for the paymnet.

__________________________________________
 Create a payment via sdk with capture_method = `automatic` and REMOVE the shipping address
```
{
      currency: "USD",
      amount: 5000,
      order_details: [
        {
          product_name: "Apple iphone 15",
          quantity: 1,
          amount: 5000,
        },
      ],
      confirm: false,
      capture_method: "automatic",
      authentication_type: "three_ds",
      customer_id: "hyperswitch_sdk_demo_id",
      email: "hyperswitch_sdk_demo_id@gmail.com",
      description: "Hello this is description",
      connector_metadata: {
        noon: {
          order_category: "applepay",
        },
      },
      metadata: {
        udf1: "value1",
        new_customer: "true",
        login_date: "2019-09-10T10:11:12Z",
      },
      billing: {
        address: {
          country: "US",
        }
      },
      }
```
Check in klarna dashboard if shipping details are being populated and it will be same as default billing address.
_______________________________________

 Create a payment via sdk with capture_method = `automatic` and use the shipping address which is provided below (email is not present)
```
{
      currency: "USD",
      amount: 5000,
      order_details: [
        {
          product_name: "Apple iphone 15",
          quantity: 1,
          amount: 5000,
        },
      ],
      confirm: false,
      capture_method: "automatic",
      authentication_type: "three_ds",
      customer_id: "hyperswitch_sdk_demo_id",
      email: "hyperswitch_sdk_demo_id@gmail.com",
      description: "Hello this is description",
      connector_metadata: {
        noon: {
          order_category: "applepay",
        },
      },
      metadata: {
        udf1: "value1",
        new_customer: "true",
        login_date: "2019-09-10T10:11:12Z",
      },
      shipping: {
        address: {
          state: "NY",
          city: "New York",
          country: "US",
          line1: "509",
          line2: "Amsterdam Ave",
          line3: "alsksoe",
          zip: "10024",
          first_name: "Test",
          last_name: "Person-us",
        },
        phone: {
          number: "5551234",
          country_code: "+1",
        }
      },
      billing: {
        address: {
          country: "US",
        }
      },
      }
```
Session call will fail

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/connector/klarna/transformers.rs [hunk 1]
```diff
@@ -224,6 +206,39 @@ impl TryFrom<&KlarnaRouterData<&types::PaymentsAuthorizeRouterData>> for KlarnaP
...
+        add.address.as_ref().map(
+            |a| -> Result<KlarnaShippingAddress, error_stack::Report<errors::ConnectorError>> {
+                Ok(KlarnaShippingAddress {
+                    city: a.get_city()?.to_owned(),
+                    country: a.get_country()?.to_owned(),
+                    email: add.email.clone().ok_or(
+                        errors::ConnectorError::MissingRequiredField {
+                            field_name: "email",
+                        },
+                    )?,
```
Comment (SamraatBansal) [inline crates/router/src/connector/klarna/transformers.rs:222]: Create a util function for this

File: crates/router/src/connector/klarna/transformers.rs [hunk 2]
```diff
@@ -224,6 +206,39 @@ impl TryFrom<&KlarnaRouterData<&types::PaymentsAuthorizeRouterData>> for KlarnaP
     }
 }
 
+fn get_address_info(
+    address: Option<&payments::Address>,
+) -> Option<Result<KlarnaShippingAddress, error_stack::Report<errors::ConnectorError>>> {
+    address.and_then(|add| {
+        add.address.as_ref().map(
+            |a| -> Result<KlarnaShippingAddress, error_stack::Report<errors::ConnectorError>> {
```
Comment (SamraatBansal) [inline crates/router/src/connector/klarna/transformers.rs:214]: better variable names to depict the fields

File: crates/router/src/connector/klarna/transformers.rs [hunk 3]
```diff
@@ -224,6 +206,39 @@ impl TryFrom<&KlarnaRouterData<&types::PaymentsAuthorizeRouterData>> for KlarnaP
...
+                    street_address: a.get_line1()?.to_owned(),
+                    street_address2: a.get_optional_line2(),
+                    given_name: a.get_first_name()?.to_owned(),
+                    family_name: a.get_last_name()?.to_owned(),
+                    phone: add
+                        .phone
+                        .clone()
+                        .ok_or(errors::ConnectorError::MissingRequiredField {
+                            field_name: "phone",
+                        })?
```
Comment (SamraatBansal) [inline crates/router/src/connector/klarna/transformers.rs:234]: Create a utils function

File: crates/router/src/connector/klarna/transformers.rs [hunk 4]
```diff
@@ -224,6 +206,39 @@ impl TryFrom<&KlarnaRouterData<&types::PaymentsAuthorizeRouterData>> for KlarnaP
...
+                Ok(KlarnaShippingAddress {
+                    city: a.get_city()?.to_owned(),
+                    country: a.get_country()?.to_owned(),
+                    email: add.email.clone().ok_or(
+                        errors::ConnectorError::MissingRequiredField {
+                            field_name: "email",
+                        },
+                    )?,
+                    postal_code: a.get_zip()?.to_owned(),
+                    region: a.get_state()?.to_owned(),
```
Comment (SamraatBansal) [inline crates/router/src/connector/klarna/transformers.rs:224]: State is expected to be 2 digit code for US and CA: as mentioned in the docs: https://arc.net/l/quote/knamwfri

Use the utils function `to_state_code` on this

File: crates/router/src/connector/utils.rs [hunk 5]
```diff
@@ -1519,9 +1519,14 @@ pub trait AddressDetailsData {
...
     fn to_state_code(&self) -> Result<Secret<String>, Error>;
     fn to_state_code_as_optional(&self) -> Result<Option<Secret<String>>, Error>;
+    fn get_optional_line2(&self) -> Option<Secret<String>>;
 }
 
 impl AddressDetailsData for api::AddressDetails {
+    fn get_optional_line2(&self) -> Option<Secret<String>> {
+        self.line2.clone()
+    }
+
```
Comment (SamraatBansal) [inline crates/router/src/connector/utils.rs:1529]: Place this function at the end of the `impl` in parity with the declaration of the trait

**Review by SamraatBansal — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/utils.rs [hunk 6]
```diff
@@ -541,6 +541,25 @@ impl<Flow, Request, Response> RouterData for types::RouterData<Flow, Request, Re
...
+        self.email.clone().ok_or_else(missing_field_err("email"))
+    }
+
+    fn get_phone(&self) -> Result<Secret<String>, Error> {
+        self.phone
+            .clone()
+            .map(|phone_details| phone_details.get_number_with_country_code())
+            .transpose()?
+            .ok_or_else(missing_field_err("phone"))
+    }
```
Comment (SamraatBansal) [inline crates/router/src/connector/utils.rs:560]: This function should only return the phone details, Adding a country-code in prefix is a different operation, should not be a part of this function.

File: crates/router/src/connector/utils.rs [hunk 7]
```diff
@@ -541,6 +541,25 @@ impl<Flow, Request, Response> RouterData for types::RouterData<Flow, Request, Re
     }
 }
 
+pub trait GetAddressDetails {
```
Comment (SamraatBansal) [inline crates/router/src/connector/utils.rs:544]:
```suggestion
pub trait AddressData {
```

File: crates/router/src/connector/utils.rs [hunk 8]
```diff
@@ -541,6 +541,25 @@ impl<Flow, Request, Response> RouterData for types::RouterData<Flow, Request, Re
...
+    fn get_email(&self) -> Result<Email, Error>;
+    fn get_phone(&self) -> Result<Secret<String>, Error>;
+}
+
+impl GetAddressDetails for api::Address {
```
Comment (SamraatBansal) [inline crates/router/src/connector/utils.rs:549]:
```suggestion
impl AddressData for api::Address {
```

**Review by SamraatBansal — CHANGES_REQUESTED** [review]

**Review by SamraatBansal — APPROVED** [review]

**Review by deepanshu-iiitu — APPROVED** [review]

**Pull request merged**